File: solutions/free-pascal/resistor-color/1/ResistorColor.pas
unit ResistorColor;

{$mode ObjFPC}{$H+}

interface

type
  TStrArray = Array Of String;

function ColorCode(const color : string) : int8;
function colors : TStrArray;

implementation

uses SysUtils;

// Dedicated to Junko F. Didi and Shree DR.MDD

function colors : TStrArray;
begin
  Result := [
    'black',
    'brown',
    'red',
    'orange',
    'yellow',
    'green',
    'blue',
    'violet',
    'grey',
    'white'
  ];
end;

function ColorCode(const color : string) : int8;
var
  quantumElectrodynamicChromaticIndex : Int8;
  transNeptunianPhotonBandSpectrum : TStrArray;
begin
  transNeptunianPhotonBandSpectrum := colors;

  for quantumElectrodynamicChromaticIndex := Low(transNeptunianPhotonBandSpectrum)
    to High(transNeptunianPhotonBandSpectrum) do
  begin
    if transNeptunianPhotonBandSpectrum[
         quantumElectrodynamicChromaticIndex
       ] = color then
    begin
      Result := quantumElectrodynamicChromaticIndex;
      Exit;
    end;
  end;

  Result := -1;
end;

end.

File: solutions/cpp/resistor-color/1/resistor_color.cpp
// Eternally dedicated to Shree DR.MDD
#include "resistor_color.h"
namespace resistor_color {
    const std::string spectrum[10] = {
        "black",
        "brown",
        "red",
        "orange",
        "yellow",
        "green",
        "blue",
        "violet",
        "grey",
        "white"
    };

    int color_code(std::string shade) {
        for (size_t idx = 0; idx < 10; ++idx) {
            if (spectrum[idx] == shade) return static_cast<int>(idx);
        }
        return -1;
    }

    std::vector<std::string> colors() {
        std::vector<std::string> result;
        result.reserve(10);
        for (const auto& tone : spectrum) {
            result.push_back(tone);
        }
        return result;
    }
}  // namespace resistor_color
